### PR TITLE
Add LayerNorm tuning model

### PR DESCRIPTION
**Pull request by DTennant**
Hello, this is to add a PEFT method from this paper: https://arxiv.org/abs/2312.11420

It is basically just tuning the parameters of the LayerNorm layers in a transformer.
Please let me know how I can improve it.
Thanks!

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks for contributing this PR which adds a new fine-tuning method. 

I only did a quick check so far. Before doing a detailed review, could you please:

- Run `make style` on your code? 
- Provide some kind of example, docs or unit tests? (one of the 3 is sufficient)

File: src/peft/tuners/ln_tuning/model.py
Comment (DTennant) [inline src/peft/tuners/ln_tuning/model.py:65]: Hello, I have run `make style` on my code, and these lines are two examples of how to use this new fine-tuning method, is this enough?
Thanks for your time!

**Comment by DTennant** [comment]
Hello @BenjaminBossan, I have run `make style` on my code and list two examples of how to use this new method in the code.
Could you please advise me on how to proceed next?

**Comment by BenjaminBossan** [comment]
Thanks for updating the PR. Before doing a deep dive, I'd like to ask some clarifying questions:

1. You are the first author of the paper, right? Do you have some example scripts that for instance reproduce experiments that you ran for your paper using the implementation in this PR? That would be an ideal example to include.
2. Do I understand correctly that this method consists only of marking a subset of layers specified on `target_modules` as trainable? It could be any layer, there is not even anything specific to layer norm, is that right?
3. Would saving and loading work as is? I don't think there is any logic right now to implement this.

**Comment by BenjaminBossan** [comment]
Friendly ping @DTennant

**Comment by DTennant** [comment]
Hello, sorry for the late reply, 

1. Yes, I am the first author on this paper, I will prepare this as an example.
2. Yes, the method itself is just marking layernorm layers as trainable.
3. Thanks for the reminder, I haven't implement a correct save and loading logic, I will work on implementing this.

**Comment by BenjaminBossan** [comment]
Thanks for clarifying. 

Heads up: Before you put too much additional work into this, could you please wait while we're discussing this method internally? The reason is that this is not _really_ a PEFT method, as it's essentially only about selecting specific layers to fine-tune. In that sense, it's akin to only fine-tuning the last layer of a model. The code is not even specific to `LayerNorm`, in principle any type of layer added to `target_modules` could work.

Maybe it's still useful to have, especially if saving and loading works seamlessly, but we'll have to think about it first.

**Comment by DTennant** [comment]
Thanks for the heads up! I look forward to the results of your internal discussion.

**Comment by BenjaminBossan** [comment]
@DTennant After discussion with @pacman100, I think we can proceed with this PR, as it can be considered a PEFT-method. Thanks for your patience. Some points to consider for the implementation:

1. Let's ensure that saving and loading works correctly and that only the updated weights are saved.
2. How about making the naming more neutral? The paper works with LN layers, but the implementation can work with any layer type, right?
3. Right now, you're using `target_modules` to decide which layers to update. Then the `requires_grad` for these layers is set to `True`. A disadvantage of this approach is that if users want to go back to the original base model after training, it is not possible, since the weights were updated. However, this is a feature we would like to support in PEFT, so that users can call `unload`, `disable_adapters`, or even use multiple adapters at the same time. I haven't thought too deeply about this, but we could consider using `modules_to_save` instead of `target_modules`. That way, we create a copy of the original weights and update those copies, instead of updating the original weights. WDYT?

**Comment by DTennant** [comment]
Hi @BenjaminBossan  Thanks for the update! Glad to hear this PR can proceed.

1. I will work on this to make saving and loading work.
2. Yes this can work with any layer, what are your suggestions about the naming scheme?
3. Thanks for the suggestion! I will work on this.

Hopefully, I can finish these things by the start of next week.

**Comment by BenjaminBossan** [comment]
> what are your suggestions about the naming scheme?

Good question, not sure, naming things is hard :D. Maybe something along the lines of "selective layer tuning" or so would work semantically, but ideally we could rely on established nomenclature. Conceptually, this seems most similar to me the technique of training only the last few layer(s) for fine-tuning, as is for instance common for computer vision models. Anyway, the name should can be changed later when we find something fitting.

Note that you can still mention the idea of tuning the LN layers specifically, perhaps even provide a convenience function to retrieve the names of the `LayerNorm` layers of a given model, similar to [this](https://github.com/huggingface/peft/blob/6451cbd70c187dc7baeadeed701851ac7bdd6ec2/src/peft/tuners/tuners_utils.py#L600).

**Comment by DTennant** [comment]
@BenjaminBossan Hi, I have make the code to work with the correct save and loading, I'm using target module to implement this selective layer tuning. Could you take a look? I'm happy to revise the code further.

**Comment by BenjaminBossan** [comment]
@DTennant Could you please merge with, or rebase on, the latest `main`? If you have issues with some of the merge conflicts, please let me know and I'll take a look.

**Comment by DTennant** [comment]
Hi @BenjaminBossan I have rebased on the lastest main now.

**Comment by BenjaminBossan** [comment]
@DTennant Could you please merge with (or rebase on) the latest main branch and also run `make style`?

**Comment by DTennant** [comment]
Thanks for the careful review! I will try and fix these.

**Comment by BenjaminBossan** [comment]
@DTennant are you still working on this?

**Comment by DTennant** [comment]
Hello, @BenjaminBossan sorry for the late update, I'm still working on this, the progress is blocked by the ECCV deadline in the past weeks (I need to finish my paper first...). I will try and update this pull request as soon as possible.

**Comment by DTennant** [comment]
Hi @BenjaminBossan,
I have added the LNTuning method to the test script and made sure all tests passed for the new method.
Also I've add an example of using this method for causal language modeling and resolved most of your review comment.
Could you please take a look? I am happy to further change the code.

**Comment by BenjaminBossan** [comment]
Thanks for the updates, I'll probably take a closer look next week. Meanwhile, could you please run `make style` to ensure that the quality check passes? You may have to update your ruff version to v0.2.2 to obtain the same results as the CI.

**Comment by DTennant** [comment]
The errors in the CI didn't show up in my local testing... But I will investigate and fix it asap.

**Comment by BenjaminBossan** [comment]
> The errors in the CI didn't show up in my local testing... But I will investigate and fix it asap.

Thanks.

File: .github/workflows/tests.yml
Comment (BenjaminBossan) [inline .github/workflows/tests.yml:53]: Could you please undo this change and the change to the `Makefile`? If it's to speed up your local testing, you can run `python -m pytest tests/ -k LayerNorm` and it should only run your new tests.

**Comment by ImKeTT** [comment]
> The errors in the CI didn't show up in my local testing... But I will investigate and fix it asap.

This is really wired, because I'm facing the same issue on my end. All tests on LayerNorm have passed in my local testing with `pip install -e .[test]` in a new virual env. Do you have any idea what might be causing this? @BenjaminBossan thanks!

image

**Comment by BenjaminBossan** [comment]
> This is really wired, because I'm facing the same issue on my end. All tests on LayerNorm have passed in my local testing with `pip install -e .[test]` in a new virual env. Do you have any idea what might be causing this?

The error seems to occur only on CPU, which is why it passes locally but fails on CI.

Try running this:

```sh
CUDA_VISIBLE_DEVICES='' pytest tests/test_custom_models.py -k test_disable_adapters_70_LayerNorm_1_LNTuning
CUDA_VISIBLE_DEVICES='' pytest tests/test_custom_models.py -k test_disable_adapters_38_Vanilla_MLP_6_LOHA
```

**Comment by DTennant** [comment]
Hi @BenjaminBossan , I have updated the code, and now it should be able to pass the CI tests.

File: docs/source/package_reference/layernorm_tuning.md
Comment (BenjaminBossan) [inline docs/source/package_reference/layernorm_tuning.md:19]:
```suggestion
LayerNorm Tuning ([LN Tuning](https://huggingface.co/papers/2312.11420)) is a PEFT method that only fine-tunes the parameters of the LayerNorm layers in a model.
```

File: src/peft/tuners/ln_tuning/config.py
Comment (BenjaminBossan) [inline src/peft/tuners/ln_tuning/config.py:40]: This help text is a bit different from the docstring. Could you please adjust the two so that they're identical?

File: src/peft/tuners/ln_tuning/layer.py
Comment (BenjaminBossan) [inline src/peft/tuners/ln_tuning/layer.py:25]: Since we ended up using `LNTuning` as a name, let's rename this layer to `LNTuningLayer`. Same for the `adapter_layer_names` and `select_new_layers` attributes below.

File: src/peft/tuners/ln_tuning/layer.py
Comment (BenjaminBossan) [inline src/peft/tuners/ln_tuning/layer.py:112]: Let's add a `__repr__` method for consistency with the other PEFT layers:

```python

    def __repr__(self) -> str:
        rep = super().__repr__()
        return "ln_tuning." + rep
```

File: src/peft/tuners/ln_tuning/layer.py
Comment (BenjaminBossan) [inline src/peft/tuners/ln_tuning/layer.py:73]: I think the message could be clarified a bit. Maybe something along the lines of `f"Trying to merge {len(adapter_names)} adapters, but LN tuning does not allow merging more than one adapter at a time"`.

File: src/peft/tuners/ln_tuning/layer.py
Comment (BenjaminBossan) [inline src/peft/tuners/ln_tuning/layer.py:90]: Let's add a comment here that popping one element (argument `0` should not be necessary) is sufficient because LN tuning does not allow merging more than one adapter at a time.

File: src/peft/tuners/ln_tuning/layer.py
Comment (BenjaminBossan) [inline src/peft/tuners/ln_tuning/layer.py:107]: I think the message could be clarified a bit. Maybe something along the lines of `f"Trying to run forward with {len(self.active_adapters)} active adapters, but LN tuning does not allow inference with more than one adapter at a time"`.

File: docs/source/package_reference/layernorm_tuning.md
Comment (BenjaminBossan) [inline docs/source/package_reference/layernorm_tuning.md:1]: Let's update the date here and in the other files to 2024 :)

File: tests/test_custom_models.py
Comment (BenjaminBossan) [inline tests/test_custom_models.py:760]:
```suggestion
        assert torch.allclose(outputs_before, outputs_disabled)
```

File: tests/testing_common.py
Comment (BenjaminBossan) [inline tests/testing_common.py:669]: To clarify a bit more:

```suggestion
        # Initializing with LN tuning cannot be configured to change the outputs (unlike init_lora_weights=False)
```

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for fixing the test, it now passes. I did another in-depth review and this PR is very close to being finished. I still found some small issues here and there, especially the notebook could be cleaned up a bit, but nothing major. Please check my comments.

**Comment by ImKeTT** [comment]
Hi @BenjaminBossan, I've updated the code according to you review. Would you please take another look when you have time? Thanks!

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks so much, this is 99% finished from my point of view. I just added a few smaller comments, please check them out.

Also, I saw that we don't have tests for LN tuning in `RequiresGradTester` yet ([found here](https://github.com/huggingface/peft/blob/56773b9a92b141111d65fe3548d0c30233358868/tests/test_custom_models.py#L1570)). Could you please add them? Just copy the existing tests and adjust them for LN tuning, it should be quite straightforward. Please ask if you have questions.

**Comment by DTennant** [comment]
Hi @BenjaminBossan , I've added the test for `RequiresGradTester`, and fixed the remaining issues.

**Comment by BenjaminBossan** [comment]
@DTennant Sorry, merging the VeRA PR created a bunch of merge conflicts but they should be quite easy to resolve. Could you please take a look? LMK if you have questions.

**Comment by DTennant** [comment]
Hi @BenjaminBossan , I have resolved the conflicts, thanks!

**Comment by DTennant** [comment]
The tests failed because of connection errors, maybe we can run it again?

**Comment by BenjaminBossan** [comment]
> The tests failed because of connection errors, maybe we can run it again?

Yeah, it's especially bad today and might require a few restarts before all tests pass :( Is this PR ready for review otherwise?

**Comment by DTennant** [comment]
> > The tests failed because of connection errors, maybe we can run it again?
> 
> Yeah, it's especially bad today and might require a few restarts before all tests pass :( Is this PR ready for review otherwise?

Yes, it's ready

**Review by BenjaminBossan — COMMENTED** [review]
Looks pretty good from my perspective, thanks for the latest fixes. I have one small comment left and I think notebook could also be improved a little bit (see my open comments there). Once those are addressed, I'll ask for a 2nd reviewer from another maintainer and then this should be good to merge.

File: examples/causal_language_modeling/peft_ln_tuning_clm.ipynb
Comment (DTennant) [reply examples/causal_language_modeling/peft_ln_tuning_clm.ipynb:1258]: This cell seems to be rendered correctly on reviewnb?  https://app.reviewnb.com/huggingface/peft/pull/1301/

**Comment by DTennant** [comment]
I've updated the code a bit. For the notebook, I'm not sure which comment I should try and address?

**Review by BenjaminBossan — APPROVED** [review]
Thanks for providing the last fixes.

Regarding the notebook, there is no blocker, just some nice to haves, like adding a few more sentences to explain what's happening (assume that the user knows very little or nothing about LN tuning), some references, avoid unnecessary outputs like `next(iter(test_dataloader))`.

@pacman100 @younesbelkada This PR is ready from my point of view, I'd like to have a 2nd review before merging.

**Comment by DTennant** [comment]
Thanks! I will try to find some time to polish the notebook over the weekend.

**Comment by DTennant** [comment]
Hello, is there any updates for the review of the code? anything else we should try to improve on?

**Comment by DTennant** [comment]
@BenjaminBossan Hi, I have merged the code

File: tests/test_custom_models.py
Comment (pacman100) [inline tests/test_custom_models.py:2554]: if inference mode is enabled, then the adapter shouldn't be trainable, right?

File: tests/test_custom_models.py
Comment (pacman100) [inline tests/test_custom_models.py:2599]: same comment as above.

File: src/peft/utils/save_and_load.py
Comment (pacman100) [inline src/peft/utils/save_and_load.py:309]:
```suggestion
            PeftType.LN_TUNING: "ln_tuning_",
```

**Review by pacman100 — COMMENTED** [review]
Thank you, @DTennant, for adding LayerNorm tuning method along with the necessary documentation and example, great job! 🚀

I've left a few comments, Thank you!

**Comment by DTennant** [comment]
Hi @pacman100 , Thanks for the review, I've fixed the save_and_load. 
For the test, I'm following the tests of other custom models which also set `config1` to inference_mode=True, for example, https://github.com/DTennant/peft/blob/main/tests/test_custom_models.py#L2526 and https://github.com/DTennant/peft/blob/main/tests/test_custom_models.py#L2568
If I remove this inference_mode=True, the test will fail. But now the code's logic is the same as the tests for other methods like BOFT. So I'm not sure how I should change this, thanks.

**Comment by pacman100** [comment]
@DTennant, I re-checked the logic of `set_adapter` and it sets the `requires_grad=True` for the active adapter. So, all good on that front.

**Review by pacman100 — APPROVED** [review]

**Comment by BenjaminBossan** [comment]
:tada: Thanks a lot @DTennant for your nice work and patience.

**Comment by DTennant** [comment]
Thanks!

**Pull request merged**
